<a href="https://colab.research.google.com/github/lmnopasdy/58053_CS5DataDevelopment_TSDC/blob/main/58089_PrelimPS1_Sunny.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Topic02a : Prelim Problem Set I
### Dy, Sunny A.
### 58089

In [1]:
import numpy as np
import pandas as pd
import math

## Case 1


**Problem 1.a. System of Linear Equations**
>**Dot Product Form**
$$
\begin{bmatrix}0&-1&1\\\frac{1}{2}&-2&0\\-1&0&\frac{3}{7}\end{bmatrix} \cdot \begin{bmatrix}xyz\end{bmatrix} = \begin{bmatrix}\frac{1}{32}\\0\\\frac{4}{5}\end{bmatrix}
$$

In [ ]:
X = np.array([
    [0,-1,1],
    [1/2,-2,0],
    [-1,0,3/7]
], dtype=float)
Y = np.array([
    [1/32],
    [0],
    [4/5]
],dtype=float)
print(X)
print(X.shape)
print(Y)
print(Y.shape)

[[ 0.         -1.          1.        ]
 [ 0.5        -2.          0.        ]
 [-1.          0.          0.42857143]]
(3, 3)
[[0.03125]
 [0.     ]
 [0.8    ]]
(3, 1)


In [ ]:
X_inv = np.linalg.inv(X)
X_inv

array([[ 0.48, -0.24, -1.12],
       [ 0.12, -0.56, -0.28],
       [ 1.12, -0.56, -0.28]])

In [ ]:
r = X_inv @ Y
r

array([[-0.881  ],
       [-0.22025],
       [-0.189  ]])

In [ ]:
ans=X @ r
np.allclose(ans, Y)

True

Vector v = $$\begin{bmatrix}-0.881\\-0.22025\\-0.189\end{bmatrix}$$

**Problem 1.b. Linear Combination**
>>**Vectorized Form**
$$
\begin{bmatrix}cos\hat{i}&sin\hat{j}&-csc\hat{k}\end{bmatrix}\cdot\begin{bmatrix}\theta\\\theta\\2\theta\end{bmatrix}
$$

**Problem 1.c. Scenario**

In [ ]:
attendees = np.array([200,45,15,40]) 
percom = attendees/attendees.sum(axis=0)*100
print(percom)

[66.66666667 15.          5.         13.33333333]


$$
\begin{bmatrix}AttendeeType&Quantity&Percentage\\StudentAttendees&200&66.67\%\\Professionals&45&15\%\\MembersOfThePanel&15&5\%\\OrganizingCommittee&40&13.33\%\end{bmatrix}
$$

# Case 2


**Problem 2.a: Vector Magnitude**
>The magnitude of a vector can be computed using the formula:\
$$||v||=\sqrt{a_0^2 + a_1^2 + a_n^2}$$
The magnitude of a vector can be determined if its endpoint is located at origin (0,0) which is usually represented as:\
$$|\vec{v}|=\sqrt{x^2+y^2}$$\
The starting point of a vector is located at the certain point $(x_1,y_1)$ and its ending point at $(x_2,y_2)$, this can also be represented with this formula: $$|\vec{v}|=\sqrt{(x_2-x_1)^2+(y_2-y_1)^2}$$


**Problem 2.b: Angle Between Vectors**
>The two vectors are perpendicular when their dot product is equal to zero

>This means that if the $\theta = \frac{\pi}{2}$ 

$$u\cdot v = ||u||\cdot||v||\cos(\frac{\pi}{2})$$
>and it will become 
$$u\cdot v = 0$$

# Case 3

**Problem 3.a: Matrix Equivalences**

We can identify if $X \cdot \theta$ and $\theta^T\cdot X$ are equal by getting the  matrix product of $X \cdot \theta$

$$
X \theta = \begin{bmatrix} 
— (x^{(1)})^T— \\ 
— (x^{(2)})^T— \\
\vdots \\
— (x^{(m)})^T— \\
\end{bmatrix}
\cdot\begin{bmatrix} 
\theta^{(1)} \\ 
\theta^{(2)} \\
\vdots \\
\theta^{(m)} \\
\end{bmatrix} 
$$

By computing the matrix product $X\theta$ we have:

$$X\theta = 
\begin{bmatrix} 
— (x^{(1)})^T\theta— \\ 
— (x^{(2)})^T\theta— \\
\vdots \\
— (x^{(m)})^T\theta— \\ 
\end{bmatrix} = 
\begin{bmatrix} 
— \theta^T(x^{(1)})— \\ 
— \theta^T(x^{(2)})— \\
\vdots \\
— \theta^T(x^{(m)})— \\ 
\end{bmatrix}
$$
The product of $X\theta$ is also equal to its transpose which is $\theta^TX$

**Problem 3.b: Matrix Shapes**
$$X = \begin{bmatrix} 
a^1&a^2&a^3&a^4&a^5 \\ 
a^6&a^7&a^8&a^9&a^{10} \\
\vdots&\vdots&\vdots&\vdots&\vdots \\
a^{296}&a^{297}&a^{298}&a^{299}&a^{300} \\
\end{bmatrix} \text{, } 
\theta = \begin{bmatrix} 
\theta^{(1)} \\ 
\theta^{(2)} \\
\vdots \\
\theta^{(m)} \\
\end{bmatrix}$$
>The logistic regression hypothesis can be defined as:
$$h_\theta(x) = g(\theta^Tx),$$\
Function $g$ is the sigmoid function and can be defined as:
$$g(z) = \frac{1}{1+e^{-z}}$$
>The shape of $h_\theta(x)$ would be likely the same as the shape of the X that has (300,5).
$$h_\theta(x) = \begin{bmatrix} 
\theta^1(a)&\theta^2(a)&\theta^3(a)&\theta^4(a)&\theta^5(a) \\ 
\theta^6(a)&\theta^7(a)&\theta^8(a)&\theta^9(a)&\theta^{10}(a) \\
\vdots&\vdots&\vdots&\vdots&\vdots \\
\theta^{296}(a)&\theta^{297}(a)&\theta^{298}(a)&\theta^{299}(a)&\theta^{300}(a) \\
\end{bmatrix}$$

**Problem 3.c: Vectorization**

$$J(\theta) = \frac{1}{m} \sum^m_{i=0}=[-y^{(i)}\log({h_{\theta}(x^{(i)})})-(1-y^{(i)})\log(1-h_{\theta}(x^{(i)}))]\\_{\text{Eq. 3.4}}$$

The gradient vector:

$$\frac{∂J(\theta)}{∂(\theta)_0} = \frac{1}{m} \sum^m_{i=0} ({h_{\theta}(x^{(i)}})-y^{(i)})x^{(i)}
_{j}$$

$$\frac{∂J(\theta)}{∂(\theta)_j} = \frac{1}{m} \sum^m_{i=0} ({h_{\theta}(x^{(i)}})-y^{(i)})x^{(i)}
_{j}$$

Vectorized form:

$$ = \frac{1}{m} X^T({h_{\theta}(x)}-y)$$

**Problem 4.c: Computational Programming (Also Laboratory 2)**